### ISSUE TITLE
Enforce usage of QueryGraph lookup ?

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1744]
Right now we almost always do:

```java
QueryGraph queryGraph = new QueryGraph(graph);
queryGraph.lookup(queryResults);
```

I have moved usages of this pattern into a static helper method here: https://github.com/graphhopper/graphhopper/commit/287fef26750a494bfcd840f815383952d3cf3329, https://github.com/graphhopper/graphhopper/commit/ff044d10594e4a4a11d05901bf5a582259a2c618

Now there are only a few usages that do not strictly follow this pattern left and I am wondering if we can get rid of them altogether.

I think my point is that for the most part `QueryGraph` cannot be used as long as `lookup` has not been called anyway and not having the query results by the time we create the `QueryGraph` makes the initialization code more complicated. On the other hand there is some ugliness involved, because `lookup` manipulates the query results. So calling a constructor like `new QueryGraph(graph, queryResults)` would be ugly because it is already running some algorithm and manipulates `queryResults`. But honestly I am not sure if it is more ugly when the same happens once we call `lookup` (?). 
Anyway, I think the real problem here is that `QueryGraph` is even performing some algorithm (instead of just being a graph data structure). Ideally, the algorithm part (calculating the virtual edges) should be moved outside of query graph anyway (that's what I was trying to do).

Background: I need to use the virtual edge creation logic, but do not want (or at least I think I do not want) to use the (very wide) `Graph` interface... So my vision was something like this (note how the virtual edge/node calculation happens outside of the query graph):

```java
List<QueryResults> queryResults = getQueryResultsFromLocationIndexEtc();
Result result = QueryResultProcessor.calculateVirtualNodesAndEdges(queryResults, virtualNodeIdsStart, virtualEdgeIdsStart);
QueryGraph queryGraph = new QueryGraph(graph, result);
```

**Comment by easbar** [comment]
The only place where I am not sure is `GraphHopperGtfs#route` @michaz ?

**Comment by michaz** [comment]
Hmmm, not sure I can play through your proposed refactoring in my head and then suggest how to apply it to my code.

Just a thought about your snippet: I think the classes/objects involved are getting more and more artificial. I agree that the step where _everything_ is modified (`QueryGraph` _and_ `QueryResult`) is confusing, and `QueryResult` is not a good class name (what query? what result?). But a `QueryResultProcessor` that takes a `QueryResult` and produces a `Result` is over the top, especially since `Result` is not purely a result, but is an input for `QueryGraph`.

I think we should try to make every object that is passed from one important piece of code to another (like from location index to query graph) a _thing_ that can be given a _name_ that pins down what it _is_. E.g. in OpenTripPlanner, a thing similar to QueryResult is called a `Split` - which means a point on an edge.

Not sure if helpful.

**Comment by michaz** [comment]
To clarify, I think `Query`, `Result` and even `QueryResult` are totally okay names if they are _really_ just parameter types or result types, and are created immediately before use and thrown away immediately after use.

They are not okay if they are important application concepts that several pieces of code know about.

**Comment by easbar** [comment]
> But I think you can just create the QueryGraph in the line with the lookup instead of in the constructor, and it will be fine.

:+1: did that

> I think the classes/objects involved are getting more and more artificial. I agree that the step where everything is modified (QueryGraph and QueryResult) is confusing, and QueryResult is not a good class name (what query? what result?). But a QueryResultProcessor that takes a QueryResult and produces a Result is over the top, especially since Result is not purely a result, but is then an input for QueryGraph. And the "query" in QueryResult probably refers to querying the location index, but the "query" in QueryGraph probably doesn't. Otherwise it would have to be called QueryResultGraph. It's almost word salad.

Agreed, `Result` and `QueryResultProcessor` are certainly no good names, I just made them up for the example. Yes, `QueryResults` are the result of the `LocationIndex` query. Maybe `SnappedPoints/Coordinates` or just `Snaps` would be better names ? Its still a bit ugly because they are further manipulated by `QueryGraph`.

> How can creating virtual edges be removed from QueryGraph anyway? Isn't that an implementation detail of QueryGraph that doesn't make sense outside of it? A virtual edge outside of QueryGraph -- what Graph is it an Edge of? None? What is an Edge without a Graph?

`QueryGraph` does two (more or less) unrelated things: 

1) It sorts the snapped points by their closest edges and determines which virtual edges are needed to connect the virtual nodes with each other and/or the adjacent real node. It also determines which real edges no longer should be used, because instead of using them we should travel via the new virtual nodes that are located in between the real node. Basically it determines what I now called `GraphModification` (<- I still do not like that name very much) that consists of new and forbidden edges. This data 'exists without a graph' because it can be useful even without `QueryGraph`. Then again of course its related to the actual graph because it refers to it via node/edge ids.
2) It uses this information and combines it with the real graph to create a new graph (`QueryGraph` kind of constructs itself).

**Comment by easbar** [comment]
Closed via #1751.
